plugins/ingest/src/main/java/org/elasticsearch/ingest/Data.java

```diff
@@ -42,14 +45,34 @@ public Data(String index, String type, String id, Map<String, Object> document)
         this.document = document;
     }
 
+    // TODO(talevy): support elements of lists
     @SuppressWarnings("unchecked")
     public <T> T getProperty(String path) {
         return (T) XContentMapValues.extractValue(path, document);
     }
 
-    public void addField(String field, Object value) {
+    /**
+     * add `value` to path in document. If path does not exist,
+     * nested hashmaps will be put in as parent key values until
+     * leaf key name in path is reached.
+     *
+     * @param path The path within the document in dot-notation
+     * @param value The value to put in for the path key
+     */
+    public void addField(String path, Object value) {
         modified = true;
-        document.put(field, value);
+
+        String[] pathElements = Strings.splitStringToArray(path, '.');
+
+        String writeKey = pathElements[pathElements.length - 1];
+        Map<String, Object> inner = document;
+
+        for (int i = 0; i < pathElements.length - 1; i++) {
+            inner.putIfAbsent(pathElements[i], new HashMap<String, Object>());
```

Comment:
oh, no its fine. I was just in search of ways to reduce conditional branches to make things more readable. I will just go ahead and add an if to remove this java8 dep